File: t1809e/java/session/session6/src/PhoneBook.java
import java.util.ArrayList;
import java.util.Collections;

public class PhoneBook extends Phone {

    private ArrayList<PhoneNumber> PhoneList = new ArrayList<>();

    public ArrayList<PhoneNumber> getPhoneList() {
        return PhoneList;
    }

    public void setPhoneList(ArrayList<PhoneNumber> phoneList) {
        PhoneList = phoneList;
    }

    @Override
    void insertPhone(String name, String phone) {
        int result = getPhoneBookByName(name);
        if (result < 0 | result > this.PhoneList.size()) {
            this.PhoneList.add(new PhoneNumber(name, phone));
            System.out.println("Insert success");
        } else {
            PhoneNumber phoneNumber = PhoneList.get(result);
            if (phoneNumber.getPhone().equals(phone)) {
                System.out.println("Insert success");
                this.PhoneList.set(result, (new PhoneNumber(name, (phoneNumber.getPhone() + " : " + phone))));
            }
        }
    }

    @Override
    void removePhone(String name) {
        int result = getPhoneBookByName(name);
        if (result < 0 | result > this.PhoneList.size()) {
            System.err.println("Remove error!");
        } else {
            System.out.println("Remove Success");
            this.PhoneList.remove(result);
        }
    }

    @Override
    void updatePhone(String name, String newphone) {
        int result = getPhoneBookByName(name);
        if (result < 0 | result > this.PhoneList.size()) {
            System.err.println("Update error!");
        } else {
            System.out.println("Update success");
            this.PhoneList.set(result, new PhoneNumber(name, newphone));
        }
    }

    @Override
    void searchPhone(String name) {
        for (PhoneNumber phoneNumber : PhoneList) {
            if (phoneNumber.getName().equals(name)) System.out.println("Result: " + phoneNumber.getPhone());
        }
        System.out.println("Not found" + name);
    }

    @Override
    void sort() {
        Collections.sort(this.PhoneList, new NameComparator());
    }

    private int getPhoneBookByName(String name) {
        for (int i = 0; i < PhoneList.size(); i++) {
            PhoneNumber phoneNumber = PhoneList.get(i);
            if (phoneNumber.getName().equals(name)) return i;
        }
        return -1;
    }
}

File: t1809e/c#/Assignment-4-PhoneBook/PhoneBook.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;

namespace Assignment_4_PhoneBook
{
    public class PhoneBook : Phone
    {
        private List<PhoneNumber> _phoneNumbers = new List<PhoneNumber>();

        public override void InsertPhone(string name, string phone)
        {
            var index = FindByName(name);
            if (index == -1)
            {
                _phoneNumbers.Add(new PhoneNumber(name, phone));
            }
            else
            {
                var newPhone = new StringBuilder();
                newPhone.Append(_phoneNumbers[index].Phone);
                newPhone.Append(" : ");
                newPhone.Append(phone.Trim());
                _phoneNumbers[index].Phone = newPhone.ToString();
            }
        }

        public override bool RemovePhone(string name)
        {
            var index = FindByName(name);
            if (index == -1)
            {
                return false;
            }
            
            _phoneNumbers.RemoveAt(index);
            return true;
        }

        public override bool UpdatePhone(string name, string newphone)
        {
            var index = FindByName(name);
            if (index == -1)
            {
                return false;
            }

            _phoneNumbers[index].Phone = newphone;
            return true;
        }

        public override PhoneNumber SearchPhone(string name)
        {
            var index = FindByName(name);
            if (index == -1)
            {
                return null;
            }
            
            return _phoneNumbers[index];
        }

        public void Show()
        {
            Console.WriteLine("{0, 8}", "PhoneBook");
            Console.WriteLine("{0, 10} | {1, 10}", "Name", "Phone");
            foreach (var phoneNumber in _phoneNumbers)
            {
                Console.WriteLine("{0, 10} | {1, 10}", phoneNumber.Name, phoneNumber.Phone);
            }
        }

        private int FindByName(string name)
        {
            for (var i = 0; i < _phoneNumbers.Count; i++)
            {
                if (_phoneNumbers[i].Name == name)
                {
                    return i;
                }
            }

            return -1;
        }
    }
}